File: EFTAIOS/src/model/DangerousSectorCard.java
package model;
import java.util.Stack;
import java.util.Collections;


public class DangerousSectorCard{
    private CardType aType;
    private static Stack<DangerousSectorCard> drawnCards = new Stack<>();
    private static Stack<DangerousSectorCard> deckOfCards = new Stack<>();

    //this method should probably never need to be used, save for tests
    public static Stack<DangerousSectorCard> getDeckInstance(){
    	return deckOfCards;
    }
    
    static{
    	initializeDeck();
    }
    
    private static void initializeDeck(){
        for(int i = 0; i < 5; i++){
            deckOfCards.push(new DangerousSectorCard(CardType.SILENCE));
        }

        for(int i = 0; i < 10; i++){
            deckOfCards.push(new DangerousSectorCard(CardType.NOISE_IN_SECTOR));
        }

        for(int i = 0; i < 10; i++){
            deckOfCards.push(new DangerousSectorCard(CardType.NOISE_IN_OTHER_SECTOR));
        }

        Collections.shuffle(deckOfCards);
    }
    
    //used for testing
    private static void reset(){
    	deckOfCards = new Stack<>();
    	drawnCards = new Stack<>();
    	initializeDeck();
    }

    public CardType getType(){
        return aType;
    }

    private static void refillDeck(){
        for(DangerousSectorCard c : drawnCards){
            deckOfCards.push(c);
        }
        drawnCards = new Stack<>();
        Collections.shuffle(deckOfCards);
    }

    public static DangerousSectorCard draw(){
        if(deckOfCards.isEmpty()){
            refillDeck();    
        }
        DangerousSectorCard card = deckOfCards.pop();
        drawnCards.push(card);
        return card;
    }

    private DangerousSectorCard(CardType pType){
        aType = pType;
    } 

}

File: EFTAIOS/src/model/cards/DangerousSectorCard.scala
package model.cards

import model.cards._

import scala.util.Random

case class DangerousSectorCard(cardType: CardType)

object DangerousSectorCard {
  val initialDeck: Deck[DangerousSectorCard] = initializeDeck()

  private def initializeDeck(): Deck[DangerousSectorCard] = {
    val silenceCard = new DangerousSectorCard(Silence)
    val silenceCards = List.fill(5)(silenceCard)
    val noiseCard = new DangerousSectorCard(NoiseInSector)
    val noiseCards = List.fill(10)(noiseCard)
    val otherNoiseCard = new DangerousSectorCard(NoiseInOtherSector)
    val otherNoiseCards = List.fill(10)(otherNoiseCard)

    val deck = Random.shuffle(silenceCards ++ noiseCards ++ otherNoiseCards)
    Deck(deck, List())
  }
}
